feat(test toolkit): add documentation for business data (#1901)
bonitasoft/bonita-doc: modules/process-testing/pages/business-data.adoc

=== As a generic Object

Any business data can be retrieved as a generic Object using its name. The generic Object `BusinessData` gives access to the fields generically: it uses the name and the type of the field to retrieve it.

[source, Java]
----
@BonitaTests
class ProcessIT {

    @Test
    void retrieve_business_data(BonitaTestToolkit toolkit) {
        User user = toolkit.getUser("walter.bates");
        ProcessDefinition processDef = toolkit.getProcessDefinition("myProcess");
        ProcessInstance processInstance = processDef.startProcessFor(user);

        BusinessData parentObjectData = processInstance.getBusinessData("parentObject"); <1>

        BusinessData singleRef = parentObjectData.getReference("singleRef"); <2>

        Integer intField = singleRef.getIntegerField("intField"); <3>
        LocalDate localDateField = singleRef.getLocalDateField("dateOnlyField"); <4>
        List<Long> multipleLongField = singleRef.getMultipleLongField("multiLongField"); <5>
    }
}
----
<1> A business data is retrieve from the process instance as a generic object, using the business data name.

> <2> A complex field is retrieve from the generic business data: the method `getReference` is used to indicate that the field to retrieve is a reference (i.e a complex field), and the name of the field is passed to the method. Lazy and eager references are retrieved the same way.

Reviewer: ```suggestion
<2> A relation field is retrieve from the generic business data: the method `getReference` is used to indicate that the field to retrieve is a reference (i.e a relation field), and the name of the field is passed to the method. Lazy and eager references are retrieved the same way.
```

        
        interface ChildReference {
            
            long getPersistenceId();
            
            Integer getIntegerField();

            LocalDate getLocalDateField();
            
            List<Long> getMultipleLongField();
        }
    }

    @Test
    void retrieve_business_data(BonitaTestToolkit toolkit) {
        User user = toolkit.getUser("walter.bates");
        ProcessDefinition processDef = toolkit.getProcessDefinition("myProcess");
        ProcessInstance processInstance = processDef.startProcessFor(user);

        ParentObject typedParentObjectData = processInstance.getBusinessData("parentObject", ParentObject.class); <2>

        ChildReference typedSingleRef = typedParentObjectData.getSingleRef(); <3>

        Integer intField = typedSingleRef.getIntegerField(); <4>
        LocalDate localDateField = typedSingleRef.getLocalDateField();
        List<Long> multipleLongField = typedSingleRef.getMultipleLongField();
    }

}
----
<1> An interface matching the Business Data Model is created.
<2> A business data is retrieved from the process instance using its name and the type of the business data -> An object of type `ParentObject` is returned.

> <3> The complex field `singleRef` is retrieve using the Java method `getSingleRef()`. An object of type `ChildReference` is returned.

Reviewer: ```suggestion
<3> The relation field `singleRef` is retrieved using the Java method `getSingleRef()`. An object of type `ChildReference` is returned.
```

The difference between using a generic or a typed dao is the same as for business data retrieved from process instances: +
While the generic option doesn't require any development and is fully functional, the typed option requires the development of a Java model matching the business data model in order to manipulate strongly typed objects when using Business data.

=== Using a generic DAO

A generic DAO can be created at any time from the bonita test toolkit. It requires the fully qualified name of the associated business object, and then gives access to the query generically. The business data returned by the queries are also generic.

[source, Java]
----
@BonitaTests
class ProcessIT {

    @Test
    void retrieve_business_data_through_dao(BonitaTestToolkit toolkit) {
        BusinessObjectDAO<BusinessData> parentObjectDAO = toolkit.getBusinessObjectDAO("com.company.model.MyObject"); <1>

        List<BusinessData> findResult = parentObject.find(0, 100));<2>

>         BusinessData singleQueryResult = parentObject.querySingle("singleQueryName", List.of("parameterName=value"))); <3>

Reviewer: Update with latest changes

>         List<BusinessData> multipleQueryResult = parentObject.query("multipleQueryName", List.of("parameterName=value"), 0, 100)); <4>

Reviewer: Update with latest changes

>         Integer countForFind = parentObject.querySingle("countForFind", List.of(), Integer.class)); <5>

Reviewer: Update with latest changes
 (remove empty parameters list)

<5> Call the single query `countForFind`. This query is a _count_, it counts the number of values that would be returned by the associated query. The return type (`Integer`) has to be passed to the generic DAO for count queries.
<6> If the return type is not excplicitly set for a count query an error is thrown.

=== Using a typed DAO

A typed DAO works in the exact same way as a generic DAO, but the returned objects are strongly typed. +
A typed DAO can be created at any time from the bonita test toolkit. It requires the type and the fully qualified name of the associated business object.

[source, Java]
----
@BonitaTests
class ProcessIT {

    interface MyObject { <1>
            
        long getPersistenceId();
        
        Integer getIntegerField();

        LocalDate getLocalDateField();
        
        List<Long> getMultipleLongField();
    }

    @Test
    void retrieve_business_data_through_dao(BonitaTestToolkit toolkit) {
        BusinessObjectDAO<MyObject> parentObjectDAO = toolkit.getBusinessObjectDAO("com.company.model.MyObject", MyObject.class); <2>

        List<MyObject> findResult = parentObject.find(0, 100));<3>

>         MyObject singleQueryResult = parentObject.querySingle("singleQueryName", List.of("parameterName=value"))); <4>

Reviewer: Update with latest changes

}
----

> <1> An Interface matching the Business Data Model is created.

Reviewer: ```suggestion
<1> An interface matching the Business Data Model is created.
```

The Bonita test toolkit offers the possibility to create a _DAO (Data Access Object)_ for all business objects defined in the business data model. These DAO can be generic or typed, and give the possibility to retrieve business data from the BDM database using _queries_ (default or custom). 

The difference between using a generic or a typed dao is the same as for business data retrieved from process instances: +

> Whereas the generic option doesn't require any development and is fully functional, the typed option requires the development of a Java model matching the business data model, but objects returned by the queries will be correctly typed.

Reviewer: ```suggestion
While the generic option doesn't require any development and is fully functional, the typed option requires the development of a Java model matching the business data model in order to manipulate strongly typed objects when using Business data.
```

=== Using a typed DAO

> A typed DAO works in the exact same way as a generic DAO, but the returned objects are typed correctly. +

Reviewer: ```suggestion
A typed DAO works in the exact same way as a generic DAO, but the returned objects are strongly typed. +
```

    }
}
----

> <1> An Interface matching the Business Data Model is created.

Reviewer: ```suggestion
<1> An interface matching the Business Data Model is created.
```

<2> A typed DAO is created using the fully qualified name of the business object and the type of the object in the test Java model.

> <3> The query find is called, a list of `MyObject` is returned.

Reviewer: ```suggestion
<3> The `find` query is executed and a list of `MyObject` is returned.
```
